### ISSUE TITLE
Find references for shorthand properties

### ISSUE THREAD

**Issue by JsonFreeman** [microsoft/TypeScript#1507]
``` ts
interface A {
    foo();
}

function foo() { }
var a: A = { foo };
a.foo(); // find references on call to foo
```

If you find references on the call to foo, it should also turn up the actual function definition. But the function definition seems to be excluded.

**Comment by mhegazy** [comment]
I do not think this is intuitive. this quickly becomes find string. The current behavior will select all references of "foo" if you start your search by the shorthand property, but not if you start by the value, or the property name. i believe this is reasonable.

**Comment by CyrusNajmabadi** [comment]
Reactivating because this experience is really unpleasant today.  With our own coding patterns we're constantly in the state where we do a find references, then have to go to the place where the object literal is created, then do a goto-def on the shorthand property to map to the impl.

Find-references in C#/VB has the concept of 'cascading' where they will naturally jump to finding references to another symbol in cases like this.  In order to keep the results manageable, and to not make people go through results they may not actually want, the references are bucketed by definition.

So, in @JsonFreeman's example, there would be results that contained two definitions, with appropriate references to each definition contained underneath.

**Comment by JsonFreeman** [comment]
I agree, find-refs should be transitive.

**Comment by DanielRosenwasser** [comment]
:+1: let's at least experiment with this.

> this quickly becomes find string.

I don't think this is necessarily the case, and even if it is, it would be convenient not to have to differentiate if the experience is indeed decent.

**Comment by JsonFreeman** [comment]
Now that we have the ability to group by declarations, it should not be so noisy to get lots of results

**Comment by OliverJAsh** [comment]
I think I found an issue that is related to this, but slightly different. It also involves usage of object property value shorthand.

``` ts
type StateProps = {
  foo: number; // find all references of `foo` does not show usage in property value shorthand
};

declare const stateProps: StateProps;

const { foo } = stateProps;

const bar = {
  foo // expected foo to appear as reference
};
```

![image](https://user-images.githubusercontent.com/921609/36911150-9328a0f2-1e3a-11e8-8632-19b2ec3349db.png)

![image](https://user-images.githubusercontent.com/921609/36911190-ae23fcb2-1e3a-11e8-9987-992aa301ee5a.png)

**Comment by RyanCavanaugh** [comment]
Closing language service bugs related to the 6.0 implementation. For more information, see #62827
